lib/api/apiUtils/authorization/permissionChecks.js

```diff
@@ -268,75 +268,120 @@ function checkBucketPolicy(policy, requestType, canonicalID, arn, bucketOwner, l
+                results[_requestType] = actionImplicitDenies[_requestType] === false;
+                return;
+            }
+        } else {
+            if (!objectMD) {
+                // User is already authorized on the bucket for FULL_CONTROL or WRITE or
+                // bucket has canned ACL public-read-write
+                if (parsedMethodName === 'objectPut' || parsedMethodName === 'objectDelete') {
+                    results[_requestType] = actionImplicitDenies[_requestType] === false;
+                    return;
+                }
+                // check bucket has read access
+                // 'bucketGet' covers listObjects and listMultipartUploads, bucket read actions
+                results[_requestType] = isAuthorized('isBucketAuthorized', bucket, null, 'bucketGet', canonicalID,
+                    authInfo, log, request, false);
+                return;
+            }
+            if (objectMD['owner-id'] === canonicalID && requesterIsNotUser) {
+                results[_requestType] = actionImplicitDenies[_requestType] === false;
+                return;
+            }
+            // account is authorized if:
+            // - requesttype is included in bucketOwnerActions and
+            // - account is the bucket owner
+            // - requester is account, not user
+            if (bucketOwnerActions.includes(parsedMethodName)
+            && (bucketOwner === canonicalID)
+            && requesterIsNotUser) {
+                results[_requestType] = actionImplicitDenies[_requestType] === false;
+                return;
+            }
+        }
+        const aclPermission = method === 'isBucketAuthorized' ?
+            checkBucketAcls(bucket, _requestType, canonicalID, mainApiCall) :
+            checkObjectAcls(bucket, objectMD, parsedMethodName, canonicalID,
+                requesterIsNotUser, isUserUnauthenticated, mainApiCall);
+        const bucketPolicy = bucket.getBucketPolicy();
+        if (!bucketPolicy) {
+            results[_requestType] = withIam === true ? actionImplicitDenies[_requestType] === false :
+            actionImplicitDenies[_requestType] === false && aclPermission;
+            return;
+        }
+        const bucketPolicyPermission = checkBucketPolicy(bucketPolicy, _requestType, canonicalID,
+            arn, bucketOwner, log, request);
+
+        if (bucketPolicyPermission === 'explicitDeny') {
+            results[_requestType] = false;
+            return;
+        }
+        // If the bucket policy returns an allow, we accept the request, as the
+        // IAM response here is either Allow or implicit deny.
+        if (bucketPolicyPermission === 'allow') {
+            results[_requestType] = true;
+            return;
+        }
+        // eslint-disable-next-line max-len
+        results[_requestType] = withIam === true ? actionImplicitDenies[_requestType] === false :
+            actionImplicitDenies[_requestType] === false && aclPermission;
```

Comment:
I find it hard to understand what's going on compared to the old code, i suggest using more helper functions that separate the logic and that don't depend on the caller method